File: labosi/Primjeri/Visinska mapa/mapa.cpp
/* Sveučilište u Zagrebu, Fakultet elektrotehnike i računarstva
   Umjetna inteligencija
 
   Pretraživanje prostora stanja: Pronalaženje puta na visinskoj mapi
   v1.0
 
   Copyright: (c) 2011 Frane Šarić <frane.saric@fer.hr>
 
   Zadatak:

   Napišite program koji će korištenjem algoritama A*, pretraživanja s
   jednolikom cijenom i pohlepnim pretraživanjem "najbolji prvi"
   izračunavati najjeftiniji put između dvije točke na zadanoj visinskoj
   mapi. Mapa se sastoji od diskretnih polja poredanih u m redaka i n
   stupaca. Polje na koordinatama (redak, stupac) nalazi se na visini
   v(redak, stupac). Dozvoljen je pomak od polja na koordinatama (r1,s1) na
   polje na koordinatama (r2,s2) ukoliko je |r1-r2|<=1,|s1-s2|<=1 i
   abs(v(r1,s1)-v(r2,s2))<=maks_skok. Cijena puta od polja (r1,s1) do polja
   (r2,s2) je
   sqrt((r2-r1)^2+(s2-s1)^2)+(sgn(v(r2,s2)-v(r1,s1))/2+1)*abs(v(r1,s1)-v(r2,s2)).
   Ovime se ostvaruje da je cijena uspinjanja nešto veća od cijene silaska
   (funkcija sgn vraća -1, 0 ili 1 ovisno o tome je li argument manji,
   jednak ili veći od nule). Mapu i parametre program treba učitati iz
   zadane tekstne datoteke. Program treba u grafičkome sučelju (ili na
   generiranoj slici) prikazati visinsku mapu, sve zatvorene čvorove, sve
   otvorene čvorove, pronađeni put, duljinu pronađenog puta i broj koraka
   algoritma. Potrebno je osmisliti barem tri različite heuristike i
   isprobati rad algoritama s tim heuristikama.
 
   Primjer ulazne datoteke:
   http://www.fer.hr/_download/repository/ui_lab_zad1-4_ulaz%5B1%5D.txt

   Primjer prikaza visinske mape i rješenja:
   http://www.fer.hr/_download/repository/ui_lab_zad1-4_mapa.png
   http://www.fer.hr/_download/repository/ui_lab_zad1-4_rezultat.png
*/
#include <cstdio>
#include <vector>
#include <algorithm>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <cmath>

using namespace std;

int main() {
    // Učitajmo datoteku (nije podržano učitavanje datoteka sa znakom #).
    int h, w, diffmax;
    int sy, sx, ty, tx;
    scanf("%d%d%d%d%d%d%d", &h, &w, &diffmax, &sy, &sx, &ty, &tx);
    w += 2; // Mapu širimo za jedno polje sa svake strane.
    h += 2; // Polja na rubu mape odmah će biti označena kao nedostupna
    vector<int> vmap(w * h); // visine
    vector<char> closed(w * h, 1); // podatak je li čvor zatvoren ili ne
    vector<double> f_map(w * h, 1e7); // funkcija f

    auto coords = [w](int i, int j) { return i + 1 + (j + 1) * w; };

    int vmax = 0; // najveća visina (treba nam za generiranje slike)
    for (int j = 0; j < h - 2; ++j)
        for (int i = 0; i < w - 2; ++i) {
            scanf("%d", &vmap[coords(i, j)]);
            vmax = max(vmap[coords(i, j)], vmax);
            closed[coords(i, j)] = 0;
        }

    struct node { // Za svaki čvor pamtimo vrijednost funkcije g
        const double g, &f; // i pokazivač na vrijednost funkcije f
        bool operator <(const node &b) const { // poredak čvorova
            return f != b.f ? f < b.f : &f < &b.f;
        }
    };

    int trgidx = coords(tx, ty); // cilj
    // koristimo strukturu set jer podržava pronalaženje najmanjeg čvora u
    // O(log N) (podržava brisanje i dodavanje isto u O(log N)).
    set<node> open({ { 0, f_map[coords(sx, sy)] } });
    // polje backtrack koristimo za rekonstrukciju puta
    // Pristup i promjena elementa u strukturi unordered_map je O(1).
    unordered_map<int, int> backtrack;

    int num_steps = 0; // broj koraka algoritma
    while (!open.empty()) { // dok je skup otvorenih čvorova neprazan
        node cur = *open.begin(); // uzmi najmanji element u O(log N).
        open.erase(open.begin()); // izbriši ga
        int idx = &cur.f - &f_map[0]; // izračunaj indeks čvora
        closed[idx] = true; // stavi ga u listu zatvorenih

        ++num_steps;
        if (idx == trgidx) // pronađen je cilj
            break; // gotovi smo jer je heuristika optimistična i monotona

        for (int dy = -1; dy <= 1; ++dy) { // pokušaj otvoriti svako
            for (int dx = -1; dx <= 1; ++dx) { // susjedno polje
                int new_idx = idx + dx + dy * w; // novi čvor
                if (closed[new_idx]) // preskoči zatvoreno polje
                    continue;
                double new_g = cur.g + hypot(dx, dy); // izračunaj novi g
                int hdiff = vmap[new_idx] - vmap[idx];
                if (hdiff >= 0) {
                    new_g += 1.5 * hdiff;
                    if (hdiff > diffmax)
                        continue;
                } else {
                    new_g -= 0.5 * hdiff;
                    if (-hdiff > diffmax)
                        continue;
                }
                double new_f = new_g; // počni računati f = g + h
                // uračunaj zračnu udaljenost
                new_f += hypot(tx - new_idx % w + 1, ty - new_idx / w + 1);
                // zatim dodaj visinsku cijenu (pazeći na smjer)
                int diff2 = vmap[trgidx] - vmap[new_idx];
                new_f += diff2 >= 0 ? diff2 * 1.5 : -diff2 * 0.5;
                if (new_f < f_map[new_idx]) { // smanjili smo f
                    // pobriši čvor iz liste otvorenih u O(log N) vremenu.
                    open.erase(node({ 0, f_map[new_idx] }));
                    f_map[new_idx] = new_f; // promijeni vrijednost f
                    backtrack[new_idx] = idx; // zapamti odakle smo došli
                    // dodaj čvor u listu otvorenih u O(log N) vremenu.
                    open.insert(node({ new_g, f_map[new_idx] }));
                }
            }
        }
    }
    fprintf(stderr, "Cijena: %.2f\n", f_map[coords(tx, ty)]);
    fprintf(stderr, "Broj koraka: %d\n", num_steps);

    unordered_set<int> solution({ trgidx }); // rekonstruiraj rješenje
    while (backtrack.count(trgidx)) {
        trgidx = backtrack[trgidx];
        solution.insert(trgidx);
    }
    printf("P3 %d %d 255\n", w - 2, h - 2); // zapiši sliku u PNM formatu
    for (int j = 0; j < h - 2; ++j) {
        for (int i = 0; i < w - 2; ++i) {
            float h = vmap[coords(i, j)] / (float) vmax;
            float vis = closed[coords(i, j)] * .5;
            int r = (h * (1 - vis) + vis * .5) * 255;
            int g = (h * (1 - vis) + vis * .5) * 255;
            int b = (h * (1 - vis) + vis * .1) * 255;
            if (solution.count(coords(i, j)))
                r = 255, g = 0, b = 0;
            printf("%d %d %d\n", r, g, b); // pišemo trojku R G B
        }
    }
}

File: labosi/Primjeri/Visinska mapa/mapa.py
# coding=utf-8
# Sveučilište u Zagrebu, Fakultet elektrotehnike i računarstva
# Umjetna inteligencija
#
# Pretraživanje prostora stanja: Pronalaženje puta na visinskoj mapi
# v1.0
#
# Copyright: (c) 2011 Frane Šarić <frane.saric@fer.hr>
#
# Zadatak:
#
# Napišite program koji će korištenjem algoritama A*, pretraživanja s
# jednolikom cijenom i pohlepnim pretraživanjem "najbolji prvi"
# izračunavati najjeftiniji put između dvije točke na zadanoj visinskoj
# mapi. Mapa se sastoji od diskretnih polja poredanih u m redaka i n
# stupaca. Polje na koordinatama (redak, stupac) nalazi se na visini
# v(redak, stupac). Dozvoljen je pomak od polja na koordinatama (r1,s1) na
# polje na koordinatama (r2,s2) ukoliko je |r1-r2|<=1,|s1-s2|<=1 i
# abs(v(r1,s1)-v(r2,s2))<=maks_skok. Cijena puta od polja (r1,s1) do polja
# (r2,s2) je
# sqrt((r2-r1)^2+(s2-s1)^2)+(sgn(v(r2,s2)-v(r1,s1))/2+1)*abs(v(r1,s1)-v(r2,s2)).
# Ovime se ostvaruje da je cijena uspinjanja nešto veća od cijene silaska
# (funkcija sgn vraća -1, 0 ili 1 ovisno o tome je li argument manji,
# jednak ili veći od nule). Mapu i parametre program treba učitati iz
# zadane tekstne datoteke. Program treba u grafičkome sučelju (ili na
# generiranoj slici) prikazati visinsku mapu, sve zatvorene čvorove, sve
# otvorene čvorove, pronađeni put, duljinu pronađenog puta i broj koraka
# algoritma. Potrebno je osmisliti barem tri različite heuristike i
# isprobati rad algoritama s tim heuristikama.
#
# Primjer ulazne datoteke:
# http://www.fer.hr/_download/repository/ui_lab_zad1-4_ulaz%5B1%5D.txt
#
# Primjer prikaza visinske mape i rješenja:
# http://www.fer.hr/_download/repository/ui_lab_zad1-4_mapa.png
# http://www.fer.hr/_download/repository/ui_lab_zad1-4_rezultat.png

import sys
import numpy
import math
from heapq import *

def load_map(path):
    data = map(int, open(path).read().split())
    (height, width) = data[:2]
    diffmax = data[2]
    src = tuple(data[3:5])
    trg = tuple(data[5:7])
    array = numpy.array(data[7:]).reshape((height, width))
    return array, src, trg, diffmax

def astar(m, src, trg, diffmax):
    in_bounds = numpy.zeros((m.shape[0] + 2, m.shape[1] + 2), dtype=numpy.int8)
    in_bounds[1:-1, 1:-1] = 1

    opened = numpy.zeros(m.shape, dtype=numpy.int32)
    closed = numpy.zeros(m.shape, dtype=numpy.int32)
    backtrack = {}

    hdists = numpy.zeros(m.shape, dtype=numpy.float32)

    row_dist, col_dist = numpy.mgrid[0:m.shape[0], 0:m.shape[1]]
    # zračna udaljenost
    h = numpy.hypot(row_dist - trg[0], col_dist - trg[1])

    # slabija visinska heuristika
    #h += numpy.abs(m - m[trg]) * 0.5

    # stroža visinska heuristika
    h += numpy.abs(m - m[trg]) * (numpy.sign(m[trg] - m) * 0.5 + 1)

    # bez heuristike
    h[:, :] = 0

    open = []
    # sporija varijanta:
    # open.append((h[src], 0, src))

    # brža varijanta:
    heappush(open, (h[src], 0, src))

    steps = 0
    while len(open) > 0:
        # sporija varijanta:
        # minelem = min(open)
        # open.remove(minelem)
        # hval, dist, coord = minelem

        # brža varijanta:
        hval, dist, coord = heappop(open)

        if closed[coord]:
            continue

        closed[coord] = 1
        steps += 1

        if coord == trg:
            break

        h1 = m[coord]
        for dy in xrange(-1, 2):
            for dx in xrange(-1, 2):
                ncoord = (coord[0] + dy, coord[1] + dx)
                if not in_bounds[ncoord[0] + 1, ncoord[1] + 1]:
                    continue
                if closed[ncoord]:
                    continue
                h2 = m[ncoord]
                hdist = abs(h2 - h1)
                if hdist > diffmax:
                    continue
                dcost = math.hypot(dx, dy)
                if h2 > h1:
                    dcost += hdist * 1.5
                else:
                    dcost += hdist * 0.5
                ndist = dist + dcost
                hnew = h[ncoord] + ndist
                if not opened[ncoord] or hdists[ncoord] > hnew:
                    opened[ncoord] = 1
                    hdists[ncoord] = hnew

                    # sporija varijanta:
                    # open.append((hnew, ndist, ncoord))
                    # brža varijanta:
                    heappush(open, (hnew, ndist, ncoord))

                    backtrack[ncoord] = coord
    path = numpy.zeros(m.shape, dtype=numpy.int32)
    last_coord = trg
    while last_coord in backtrack:
        path[last_coord] = 1
        last_coord = backtrack[last_coord]

    return dist, path, closed, h, steps

m, src, trg, diffmax = load_map(sys.argv[1])

pathcost, path, closed, h, steps = astar(m, src, trg, diffmax)

print "Cijena puta: %.2f" % pathcost
print "Broj koraka: %d" % steps

import Image, ImageDraw

def to_image(m):
    im = Image.new("L", (m.shape[1], m.shape[0]))
    a = im.load()
    for r in xrange(m.shape[0]):
        for c in xrange(m.shape[1]):
            a[c, r] = m[r, c]
    return im

im_h = to_image(h * 255 / max(h.max(), 1))
im_m = to_image(m)

im_closed = to_image(closed * 255)
im_path = to_image(path * 255)

final = Image.new("RGB", (m.shape[1], m.shape[0]))
import ImageOps, ImageChops
final = ImageChops.composite(
            ImageOps.colorize(im_closed, (0, 0, 0), (200, 200, 100)),
            ImageOps.colorize(im_m, (0, 0, 0), (255, 255, 255)),
            to_image(closed * 128)
            )
final.paste(ImageOps.colorize(im_path, (0, 0, 0), (255, 0, 0)),
            None,
            im_path)
final.save("slika.png", "PNG")
im_m.save("mapa.png", "PNG")
im_h.save("heuristika.png", "PNG")
